core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1899,6 +1899,54 @@ object Stream {
     def foldMonoid(implicit O: Monoid[O]): Stream[F, O] =
       self.fold(O.empty)(O.combine)
 
+    def groupWithin(n: Int, d: FiniteDuration)(
+        implicit F: Effect[F],
+        timer: Timer[F],
+        ec: ExecutionContext): Stream[F, Segment[O, Unit]] = {
+      type Tick = Int
+
+      Stream.eval(async.unboundedQueue[F, Option[Either[Tick, O]]]).flatMap { q =>
+        def startTimeout(i: Tick): Stream[F, Nothing] =
+          Stream.eval {
+            async.fork {
+              timer.sleep(d) *> q.enqueue1(i.asLeft.some)
+            }
+          }.drain
+
+        def producer = self.map(Right(_)).noneTerminate.to(q.enqueue)
```

Comment:
`noneTerminate` here is not enough, we need `onFinalize(enqueue(None))`. 
If `producer` is interrupted, `concurrently` won't interrupt the consumer (since producer is on the right side), and `None` won't be enqueued. In fact we might need to introduce interruption, which is probably another sign that `concurrently` is not the right choice here, as per @changlinli 's comment below